File: python3/insertion_sort.py
#takes an array of comparable items and sorts them in place

def insertion_sort(A):
    for j in range(1,len(A)):
        key = A[j]
        #the elements in the sub-array A[0..j-1] are already sorted
        #by inserting A[j] into the subarray in the proper place
        #the sub-array A[0..j] maintains sorted order
        i = j-1
        while i > -1 and A[i] > key:
            #shift the elements in the array to the right
            #until a space is made for key
            A[i+1] = A[i]
            i -= 1
        A[i+1] = key

File: perl/insertion_sort.pl
#/usr/bin/perl
use strict;
use warnings;

#takes an array of comparable items
#returns sorted permutation of the array
sub insertion_sort {
    my @array = @_;
    #loop through n-1 elements of array
    for(my $j = 1; $j < @array; $j++) {
        my $key = $array[$j];
        my $i = $j -1;
        #shift the elements in the array to make room for $key
        while ($i >= 0 && $array[$i] > $key){
            $array[$i+1] = $array[$i];
            $i--;
        }
        #put $key in its correct position in the array
        $array[$i+1] = $key;
    }
    return @array;
}
